doc/source/dev/build_a_job_runner.rst

```diff
@@ -0,0 +1,328 @@
+
+::
+
+    self._init_monitor_thread()
+    self._init_worker_threads()
+
+2. queue\_job method - STAGE 2
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Input params: \* job\_wrapper (Object of
+`galaxy.jobs.JobWrapper <https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/__init__.py#L743>`__)
+
+Output params: None
+
+galaxy.jobs.JobWrapper is a Wrapper around 'model.Job' with convenience
+methods for running processes and state management.
+
+-  Functioning of queue\_job method. A. prepare\_job() method is invoked
+   to do some sanity checks that all runners' queue\_job() methods are
+   likely to want to do and also to build runner command line for that
+   job. Initial state and configuration of the job are set and every
+   data is associated with **job\_wrapper**.
+
+B. Submit job to the external runner and return the jobid. Accessing
+jobs data (tool submitted in Galaxy webframework) is purely from
+job\_wrapper. eg: job\_wrapper.get\_state() -> gives state of a job
+(queued/running/failed/success/...)
+
+Let us look at a means of accessing external runner's configuration
+present under destination tag of job\_conf.xml in the above example.
+
+::
+
+    job_destination = job_wrapper.job_destination
+    docker_cpu = int(job_destination.params["docker_cpu"])
+    docker_ram = int(job_destination.params["docker_memory"])
+
+A special case: User Story: A docker based external runner is present. A
+default docker image for execution is set in job\_conf.xml. A tool can
+also specify the docker image for its execution. Specification in tool
+is given more priority than the default specification. To achieve such a
+functionality. We can use the following statement:
+
+::
+
+    docker_image = self._find_container(job_wrapper).container_id
+
+Note: This pre-written method is only for getting the external
+image/container/os..
+
+C. After successful submission of job in the external runner, submit the
+job to Galaxy framework. To do that,make an object of
+AsynchronousJobState and put it in monitor\_queue.
+
+::
+
+    ajs = AsynchronousJobState(files_dir=job_wrapper.working_directory, job_wrapper=job_wrapper, job_id=job_id, job_destination=job_destination)
+    self.monitor_queue.put(ajs)
+
+3. check\_watched\_item method - STAGE 3
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Input params: \* job\_state (Object of
+`galaxy.jobs.runners.AsynchronousJobState <https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/runners/__init__.py#L400>`__)
+
+Output params: AsynchronousJobState object
+
+| Without going into much detail, assume there is a queue to track the
+  status of every job. eg:
+| |image0|
+
+The galaxy framework updates status of a job by iterating through the
+queue. During iteration, it calls check\_watched\_item with the job.
+Your responsibility will be to get status of execution of job from
+external runner and return the updated status of the job, and also to
+copy output files for completed jobs.
```

Comment:
Missing articles: "the status of a job", "During the iteration", "to get the status of execution of the job from the external runner", "copy the output files for the completed jobs".